crates/miden-lib/asm/miden/contracts/auth/multisig.masm

```diff
@@ -1,24 +1,208 @@
 use.miden::account
 use.miden::tx
+use.miden::auth
+
 use.std::crypto::dsa::rpo_falcon512
+use.std::crypto::hashes::rpo
+use.std::sys
 
 # CONSTANTS
 # =================================================================================================
 
 # Event to place the falcon signature of a provided message and public key on the advice stack.
-const.FALCON_SIG_TO_STACK=131087
+const.AUTH_REQUEST=131087
+const.UNAUTHORIZED_EVENT=131100
+
+const.THRESHOLD_SLOT=0
+const.NUMBER_OF_APPROVERS_SLOT=1
+const.PUBLIC_KEYS_MAP_SLOT=2
+const.EXECUTED_TXS_SLOT=3
+
+# CONSTANTS
+# AdviceMap Keys
+
+# Runtime Memory Addresses
+const.LOOP_ITER=100
+const.TEMP_PK=108
+const.MESSAGE=112
+const.THRESHOLD_VALUE=116
+const.OWNER_PUB_KEY=120
+
+# ERRORS
+const.NON_CONSTANT_MSG="NON CONSTANT SIGNATURES"
+const.INVALID_SIGNATURE="SIG DOES NOT MATCH DIGEST"
+
+# => []
+
+proc.verify_signatures_loop
+    push.0
+    # => [initial_iteration_value]
+
+    mem_store.LOOP_ITER
+    # => []
+
+    push.1
+    while.true
+        mem_load.LOOP_ITER
+        # => [iteration]
+
+        push.0.0.0
+        # => [KEY]
+
+        push.PUBLIC_KEYS_MAP_SLOT
+        # => [index, KEY]
+
+        exec.account::get_map_item
+        # => [OWNER_PUB_KEY]
+
+        mem_storew.TEMP_PK
+        # => [OWNER_PUB_KEY]
+
+        padw mem_loadw.MESSAGE
+        # => [MESSAGE, OWNER_PUB_KEY]
+
+        hmerge
+        # => [SIG_KEY]
+
+        adv.has_mapkey
+        # => [SIG_KEY]
+
+        adv_push.1 not
+        # => [unauthorized, SIG_KEY]
+
+        # If signatures are non-existant the tx will fail here:
+        if.true
+            mem_loadw.MESSAGE padw
```

Comment:
with this, the message is now loaded 3x from memory. How about calling `dupw` before and just keeping it around at the end of the stack?

I think same can be done with pubkey